### ISSUE TITLE
Shouldn't `value:JsonFloat` and `value:JsonDecimal` be specified as lax types?

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-spec#1232]
**Description:**
The [spec](https://ballerina.io/spec/lang/master/#lax_static_typing) says

> There is a fixed set of type descriptors that are classified as lax: specifically json is lax, and map\<T> is lax if T is lax. The only kinds of expression for which lax typing rules are specified are field-access-expr, optional-field-access-expr and checking-expr.

Shouldn't `value:JsonFloat` and `value:JsonDecimal` also be specified as lax types?

At the moment, we don't allow

```ballerina
import ballerina/lang.value;

function fn(string str) returns error? {
    value:JsonFloat j = check str.fromJsonFloatString();
    
    json|error a = j.a; // compilation error
    int b = check j.b; // compilation error
}
```

**Comment by jclark** [comment]
Yes, they ought to be lax, but we don't currently have a mechanism for langlib to define lax types.

**Comment by jclark** [comment]
Two possibilities:

1. Say that `T & json` produces a lax type.
2. Introduce a langlib only annotation to specify laxness.

**Comment by jclark** [comment]
@MaryamZi When you've had a chance to think about, please comment on whether you are OK with 1 or still prefer 2.

**Comment by hasithaa** [comment]
I think `T & json` is a powerful feature to have. which allows users to define subtype of JSON and access it as a lax type.

```
public function main() returns error? {
    map<T & json> val = { name: "John", age: 30, data: { x : 10, y : 15} };
    T z = check val.data.z;
}

type T int|string|map<T>;
```

**Comment by MaryamZi** [comment]
Let me explain the concern I had based on a scenario where we don't have control over the type of a variable/field.

If, for example, we have a `Client` in module `mod` with a remote method `get` that returns the following record

```ballerina
public type Response record {
    string id;
    map<decimal> config; 
};
```

to use lax access with `config`, we'd need the value to be assigned to a variable of type `map<decimal> & json` first, right?

```ballerina
mod:Client cl = new;
mod:Response resp = cl->get();

map<decimal> & json config = resp.config;
decimal min = check config.min;
```

But the same can be done by just assigning to a variable of type `json` (or some other compatible lax type).

```ballerina
json config = resp.config;
decimal min = check config.min;
```

But I do agree that it's useful when we have control over the types. So +1 for approach 1.

I guess for connectors/clients (e.g., type of `config` in `Response`) we'd continue to use the non-lax type?

**Comment by jclark** [comment]
But there's no point in lax access for `map<decimal>` is there? The use case is when you want a type that's quite generic but not as generic as json.

One disadvantage of `json & T` is it doesn't handle the case where you want laxness for a type that is looser than `json` (e.g. json + tables).

**Comment by jclark** [comment]
At the moment, we say:

1. `json` is lax
2. `map<T>` is lax if `T` is lax

Rule 2 makes `map<json>` lax, which makes sense since in many cases JSON will actually be `map<json>` rather than `json`.

Following 2, we could make the rule be

- `T1 & T2` is lax if either T1 is lax or T2 is lax

Rule 2 is consistent with that since `map<json>` is equivalent to `json & map<any>`.

Another consideration is that it all feels a bit unprincipled.  What's the justification for the rule? Why don't we have a similar rule for `|`?

**Comment by dulajdilshan** [comment]
Questions related to lax typing and field access expressions.

1. Should we consider `json & readonly` as a lax type? 
2. Should we allow `json & readonly` types vars for field access even though though they are lax or not?
Ex: 

```
json & readonly x = {a: true};
json|error x = x.a
```
3. If we are allowing (2) should we allow `map<json & readonly>`  for field access?
